base/array.jl

```diff
@@ -980,12 +984,22 @@ Dict{String, Int64} with 2 entries:
 function setindex! end
 
 function setindex!(A::Array{T}, x, i::Int) where {T}
+    @inline
+    x = x isa T ? x : convert(T, x)::T
+    return _setindex!(A, x, i)
```

Comment:
@aviatesk I'd like to mark this `_setindex!` for `@propagate_inbounds` without propagating for the `convert(...)` above

Do you know of a way to do that?

Reply:
It's fine to just add `@propagate_inbounds`, since it won’t incorrectly refine the effects of `setindex!(A::Array{T}, x, i::Int)`. If a caller calls `setindex!(A::Array{T}, x, i::Int)` in an `@inbounds` context, the caller’s `:noub` should be tainted accordingly.

Reply:
Right, the effects propagation is fine, but I'm not sure I agree with the `@boundscheck` semantics

This is equivalent to whether:
```julia
@inbounds v[i] = obj_with_internal_arrays
```

should elide `@boundscheck` code in the implicit `convert`

IMO, it really shouldn't - I may have no idea what logic is in the implicit `convert` but the logic I'm trying to certify is just the `x[i]` lookup

Reply:
Are you referring to the possibility that user-defined `convert` might use `@boundscheck`? I think that’s pretty unlikely, but if a user is using `@boundscheck`, they should be aware that it could be elided by a `@propagate_inbounds` caller that propagates `@inbounds`.

Reply:
Fair enough - I do agree that would be strange to use `@boundscheck` on a `convert` and that's probably how it's been working so far

This still feels like a mis-design to me - `@propagate_inbounds` is used to propagate an assumption on the inbounded-ness of your arguments to the inbounded-ness of the arguments on a callee.

But a `set/getindex` implementation might do `x[i]` in multiple places and they may not all be an x/i that can legally assume inbounded-ness. Seems like it should really be a call-site annotation (just like normal `@inbounds` - it's just the caller-deferred version)

Reply:
In any case, I agree with you that this is an unlikely issue (convert takes no index arguments) so I'll go ahead and add the propagation